Source/MaestroBurner.cpp

```diff
@@ -0,0 +1,386 @@
+            const Box& tileBox = mfi.tilebox();
+
+            const bool use_mask = !(lev == finest_level);
+
+            const Array4<const Real> s_in_arr = s_in[lev].array(mfi);
+            const Array4<Real> s_out_arr = s_out[lev].array(mfi);
+            const Array4<const Real> p0_cart_arr = p0_cart[lev].array(mfi);
+            const Array4<const Real> source_arr = source[lev].array(mfi);
+            const Array4<const int> mask_arr = mask.array(mfi);
+
+            const auto p0_arr = p0.const_array();
+
+            ParallelFor(tileBox, [=] AMREX_GPU_DEVICE(int i, int j, int k) {
+                if (use_mask && mask_arr(i, j, k))
+                    return;  // cell is covered by finer cells
+
+                auto r = (AMREX_SPACEDIM == 2) ? j : k;
+
+                Real sdc_rhoX[NumSpec];
+                for (int n = 0; n < NumSpec; ++n) {
+                    sdc_rhoX[n] = source_arr(i, j, k, FirstSpec + n);
+                }
+                auto sdc_rhoh = source_arr(i, j, k, RhoH);
+                auto sdc_p0 = spherical ? p0_cart_arr(i, j, k) : p0_arr(lev, r);
+
+                auto rho_in = s_in_arr(i, j, k, Rho);
+                Real rhoX_in[NumSpec];
+                for (int n = 0; n < NumSpec; ++n) {
+                    rhoX_in[n] = s_in_arr(i, j, k, FirstSpec + n);
+                }
+#if NAUX_NET > 0
+                Real aux_in[NumAux];
+                for (int n = 0; n < NumAux; ++n) {
+                    aux_in[n] = s_in_arr(i, j, k, FirstAux + n);
+                }
+#endif
+                auto rhoh_in = s_in_arr(i, j, k, RhoH);
+
+                Real x_test = (ispec_threshold > 0)
+                                  ? rhoX_in[ispec_threshold] / rho_in
+                                  : 0.0;
+
+                sdc_t state_in;
+                sdc_t state_out;
+
+                Real rhoX_out[NumSpec];
+#if NAUX_NET > 0
+                Real aux_out[NumAux];
+#endif
+                Real rho_out = 0.0;
+                Real rhoh_out = 0.0;
+
+                // if the threshold species is not in the network, then we burn
+                // normally.  if it is in the network, make sure the mass
+                // fraction is above the cutoff.
+                if ((rho_in > burning_cutoff_density_lo &&
+                     rho_in < burning_cutoff_density_hi) &&
+                    (ispec_threshold < 0 ||
+                     (ispec_threshold > 0 &&
+                      x_test > burner_threshold_cutoff))) {
+                    state_in.p0 = sdc_p0;
+                    state_in.rho = rho_in;
+                    for (int n = 0; n < NumSpec; ++n) {
+                        state_in.y[n] = rhoX_in[n];
+                    }
+                    state_in.y[NumSpec] = rhoh_in;
+                    for (int n = 0; n < NumSpec; ++n) {
+                        state_in.ydot_a[n] = sdc_rhoX[n];
+                    }
+                    state_in.ydot_a[NumSpec] = sdc_rhoh;
```

Comment:
this should use the index `SENTH`

Reply:
should the line above `state_in.y[NumSpec] = rhoh_in;` also use `SENTH`?

Reply:
yes